uvicorn/server.py

```diff
@@ -312,6 +310,20 @@ async def _wait_tasks_to_complete(self) -> None:
             while self.server_state.tasks and not self.force_exit:
                 await asyncio.sleep(0.1)
 
+        # Wait for servers to close. They won't do so until all connections are
+        # closed, which we've already waited for above, so this should be quick.
+        servers_closed = asyncio.gather(
+            *[server.wait_closed() for server in self.servers]
+        )
+        # Give the servers_closed future a chance to complete so we don't
+        # spuriously log about this operation.
+        await asyncio.sleep(0.1)
+        if not servers_closed.done() and not self.force_exit:
+            msg = "Waiting for servers to close. (CTRL+C to force quit)"
+            logger.info(msg)
+            while not servers_closed.done() and not self.force_exit:
+                await asyncio.sleep(0.1)
+
```

Comment:
Can't we just do this?
```suggestion
        for server in self.servers:
            await server.wait_closed()
```

Reply:
This matches the other implementations in this function.

Reply:
The (only) problem with that way @Kludex is if something goes wrong and `wait_closed` blocks again, hitting Ctrl+C again doesn't do anything. The way I wrote it is a lot more complex but preserves not only the Ctrl+C behavior, but also, showing the message iff it takes more than 100ms or whatever--just like the other operations in this function.

Reply:
What can go wrong on `wait_closed`?

Reply:
Probably nothing as of today. But this PR is necessary precisely because `wait_closed` changed behavior.

That being said, I don't feel strongly about this at all. If you prefer the simpler lines I'm absolutely fine with it.

Reply:
Please change it.